File: blitzwing-executor/src/main/scala/com/ebay/hadoop/arrow/executor/ArrowBufferWrapper.scala
package com.ebay.hadoop.arrow.executor

import com.ebay.hadoop.arrow.executor.plan.PlannedRecordBatch.JNIBufferNode

case class ArrowBufferWrapper(executorId: Long, memoryAddress: Long, length: Long) extends AutoCloseable {
  override def close(): Unit = {
    JniExecutor.freeBuffer(executorId, memoryAddress, length)
  }
}

object ArrowBufferWrapper {
  def apply(jniBufferNode: JNIBufferNode, executorId: Long): ArrowBufferWrapper = {
    new ArrowBufferWrapper(executorId, jniBufferNode.getAddress, jniBufferNode.getLength)
  }
}

File: blitzwing-common/src/main/java/com/ebay/hadoop/blitzwing/memory/ArrowBufferWrapper.java
package com.ebay.hadoop.blitzwing.memory;

public class ArrowBufferWrapper implements AutoCloseable {
  private final MemoryManager memoryManager;
  private final long address;
  private final int length;

  public ArrowBufferWrapper(MemoryManager memoryManager, long address, int length) {
    this.memoryManager = memoryManager;
    this.address = address;
    this.length = length;
  }

  @Override
  public void close() {
    memoryManager.freeBuffer(address, length);
  }

  public long getAddress() {
    return address;
  }

  public int getLength() {
    return length;
  }
}
